File: benchmarks/matlab/bench.m
function bench(manifestPath, resultsJsonPath, logPath)
%BENCH  CToon MATLAB benchmark.
%
%   No competing MATLAB implementation exists, so this is a solo run — but
%   it follows the same methodology, corpus, and JSON output schema as
%   every other language benchmark here.
%
%   The +ctoon/ package (built by the ctoon_build_mex CMake target) must
%   already be on the MATLAB path before calling this — see
%   benchmarks/matlab/CMakeLists.txt for the addpath call that does it.
%
%   BENCH(manifestPath, resultsJsonPath, logPath) additionally writes a
%   per-file diagnostic log (which files failed, and the caught error
%   message) to logPath — one line per failure, logged only during the
%   untimed pre-pass and the first repeat of each timed phase, not all 20
%   repeats. Pass '' or omit logPath to disable this (the default).

    if nargin < 3
        logPath = '';
    end
    logFid = -1;
    if ~isempty(logPath)
        logFid = fopen(logPath, 'w');
        if logFid == -1
            fprintf('warning: could not open log file %s\n', logPath);
        end
    end

    repeats = 20;

    fid = fopen(manifestPath, 'r');
    if fid == -1
        error('bench:manifest', 'Cannot open manifest: %s', manifestPath);
    end
    lines = textscan(fid, '%s', 'Delimiter', '\n');
    fclose(fid);
    paths = lines{1};

    files = struct('path', {}, 'json', {}, 'toon', {});
    for i = 1:numel(paths)
        p = strtrim(paths{i});
        if isempty(p)
            continue;
        end
        try
            f = fopen(p, 'r');
            if f == -1
                continue;
            end
            raw = fread(f, Inf, '*char')';
            fclose(f);
        catch
            continue;
        end
        files(end+1) = struct('path', p, 'json', raw, 'toon', ''); %#ok<AGROW>
    end

    if isempty(files)
        error('bench:corpus', 'Corpus manifest is empty - nothing to benchmark.');
    end

    totalJSONBytes = 0;
    for i = 1:numel(files)
        totalJSONBytes = totalJSONBytes + numel(files(i).json);
    end

    fprintf('CToon Benchmarks — MATLAB\n');
    fprintf('Corpus: %d files, %.2f MB (JSON)\n\n', numel(files), totalJSONBytes / 1e6);
    fprintf('%-8s %-14s %12s %14s %9s %10s %12s\n', ...
        'Library', 'Operation', 'Throughput', 'Docs/sec', 'Success', '', 'Total time');

    results = {};

    % -- Untimed pre-pass --
    for i = 1:numel(files)
        try
            val = jsondecode(files(i).json);
            files(i).toon = ctoon.dumps(val);
        catch err
            logFail(logFid, 'ctoon', 'pre_pass', files(i).path, err);
        end
    end

    % -- ctoon: json_to_toon --
    tic;
    opsA = 0; bytesA = 0;
    for r = 1:repeats
        for i = 1:numel(files)
            try
                val = jsondecode(files(i).json);
                ctoon.dumps(val);
                opsA = opsA + 1;
                bytesA = bytesA + numel(files(i).json);
            catch err
                if r == 1
                    logFail(logFid, 'ctoon', 'json_to_toon', files(i).path, err);
                end
            end
        end
    end
    tA = toc;
    results{end+1} = record('ctoon', 'json_to_toon', bytesA, opsA, tA, numel(files) * repeats, repeats); %#ok<AGROW>

    % -- ctoon: toon_to_json --
    tic;
    opsB = 0; bytesB = 0;
    for r = 1:repeats
        for i = 1:numel(files)
            if isempty(files(i).toon)
                continue;
            end
            try
                val = ctoon.loads(files(i).toon);
                jsonencode(val);
                opsB = opsB + 1;
                bytesB = bytesB + numel(files(i).toon);
            catch err
                if r == 1
                    logFail(logFid, 'ctoon', 'toon_to_json', files(i).path, err);
                end
            end
        end
    end
    tB = toc;
    results{end+1} = record('ctoon', 'toon_to_json', bytesB, opsB, tB, numel(files) * repeats, repeats); %#ok<AGROW>

    % -- ctoon: roundtrip (json -> toon -> parse toon -> json, chained) --
    tic;
    opsC = 0; bytesC = 0;
    for r = 1:repeats
        for i = 1:numel(files)
            try
                val = jsondecode(files(i).json);
                toon = ctoon.dumps(val);
                val2 = ctoon.loads(toon);
                jsonencode(val2);
                opsC = opsC + 1;
                bytesC = bytesC + numel(files(i).json);
            catch err
                if r == 1
                    logFail(logFid, 'ctoon', 'roundtrip', files(i).path, err);
                end
            end
        end
    end
    tC = toc;
    results{end+1} = record('ctoon', 'roundtrip', bytesC, opsC, tC, numel(files) * repeats, repeats); %#ok<AGROW>

    write_results_json(resultsJsonPath, numel(files), totalJSONBytes, results);

    if logFid ~= -1
        fclose(logFid);
        fprintf('Debug log written to %s\n', logPath);
    end
end

function logFail(logFid, library, operation, path, err)
    if logFid == -1
        return;
    end
    fprintf(logFid, '[%s] [%s] FILE: %s ERROR: %s\n', library, operation, path, err.message);
end

function r = record(library, operation, bytes, ops, seconds, attempted, repeats)
    if ops > 0
        throughput = bytes / seconds / 1e6;
    else
        throughput = 0;
    end
    successRate = ops / attempted;
    fprintf('%-8s %-14s %9.2f MB/s %14.0f %8.0f%%  %8.4f s  (x%d reps)\n', ...
        library, operation, throughput, ops / seconds, successRate * 100, seconds, repeats);
    r = struct('library', library, 'operation', operation, ...
        'throughput_mb_s', throughput, 'docs_per_sec', ops / seconds, ...
        'success_rate', successRate, 'total_time_s', seconds);
end

function write_results_json(path, nFiles, totalBytes, results)
    out = struct('language', 'matlab', ...
        'corpus', struct('files', nFiles, 'bytes', totalBytes), ...
        'results', {results});
    fid = fopen(path, 'w');
    if fid == -1
        fprintf('warning: could not write %s\n', path);
        return;
    end
    fprintf(fid, '%s', jsonencode(out));
    fclose(fid);
    fprintf('\nResults written to %s\n', path);
end

File: benchmarks/julia/bench.jl
# CToon Julia benchmark.
#
# Unlike every other language folder here, there is no peer library to
# test alongside ctoon: a search of Julia's General registry turned up
# no TOON implementation for Julia at the time this was written. So
# this is a one-library table, same as benchmarks/matlab's.
#
# Same three operations as every other language benchmark here:
#  - json_to_toon: CToon.parse (JSON text -> native Julia value) then
#    CToon.dumps (that value -> TOON text).
#  - toon_to_json: CToon.parse_toon (TOON text -> native Julia value)
#    then CToon.to_json (that value -> JSON text).
#  - roundtrip: json_to_toon's output fed through toon_to_json, as one
#    timed operation per file (matching how every other language here
#    defines "roundtrip" -- see e.g. bench.py's own comment on this).
#
# Methodology (same across every language benchmark in this repo):
#  1. Load every file in the corpus manifest into memory (untimed).
#  2. Time each operation over `REPEATS` passes over the whole corpus.
#  3. Report throughput (MB/s of bytes actually processed by successful
#     calls only) and documents/sec.
#
# If a log file path is given (3rd CLI arg), one line per failure is
# written there — only for the first repeat, not all `REPEATS` reps.
using CToon

const REPEATS = 20

struct BenchFile
    path::String
    json::String
end

function load_corpus(manifest_path::AbstractString)
    files = BenchFile[]
    for line in eachline(manifest_path)
        isempty(line) && continue
        try
            push!(files, BenchFile(line, read(line, String)))
        catch
            continue
        end
    end
    return files
end

# Minimal hand-rolled JSON writer for the fixed, small results shape
# below — not worth pulling in a whole JSON package (JSON.jl) just to
# write a handful of floats and strings.
json_escape(s::AbstractString) = replace(s, "\\" => "\\\\", "\"" => "\\\"")
json_num(x::Real) = isfinite(x) ? string(Float64(x)) : "0"

function write_results_json(path, language, n_files, n_bytes, results)
    open(path, "w") do io
        println(io, "{")
        println(io, "  \"language\": \"$(json_escape(language))\",")
        println(io, "  \"corpus\": { \"files\": $n_files, \"bytes\": $n_bytes },")
        println(io, "  \"results\": [")
        for (i, r) in enumerate(results)
            comma = i == length(results) ? "" : ","
            println(io, "    {")
            println(io, "      \"library\": \"$(json_escape(r.library))\",")
            println(io, "      \"operation\": \"$(json_escape(r.operation))\",")
            println(io, "      \"throughput_mb_s\": $(json_num(r.throughput_mb_s)),")
            println(io, "      \"docs_per_sec\": $(json_num(r.docs_per_sec)),")
            println(io, "      \"success_rate\": $(json_num(r.success_rate)),")
            println(io, "      \"total_time_s\": $(json_num(r.total_time_s))")
            println(io, "    }$comma")
        end
        println(io, "  ]")
        println(io, "}")
    end
end

struct Result
    library::String
    operation::String
    throughput_mb_s::Float64
    docs_per_sec::Float64
    success_rate::Float64
    total_time_s::Float64
end

function record!(results, library, operation, seconds, ok, attempted, bytes)
    throughput = ok > 0 ? bytes / seconds / 1e6 : 0.0
    success_rate = attempted > 0 ? ok / attempted : 0.0
    push!(results, Result(library, operation, throughput, ok / seconds, success_rate, seconds))
    println(
        rpad(library, 8), " ", rpad(operation, 14), " ",
        lpad(round(throughput; digits=2), 9), " MB/s ",
        lpad(round(Int, ok / seconds), 14), " ",
        lpad(round(Int, success_rate * 100), 8), "%  ",
        lpad(round(seconds; digits=4), 8), " s  (x$REPEATS reps)",
    )
end

function bench_json_to_toon(files, log_fail)
    ok = 0
    bytes = 0.0
    t0 = time()
    for r in 1:REPEATS
        for f in files
            try
                CToon.dumps(CToon.parse(f.json))
                ok += 1
                bytes += sizeof(f.json)
            catch e
                r == 1 && log_fail("json_to_toon", f.path, e)
            end
        end
    end
    return time() - t0, ok, bytes
end

function bench_toon_to_json(files, log_fail)
    # Needs each file's TOON text first (same untimed pre-pass every
    # other language's benchmark does) -- ctoon is the only thing that
    # can produce that here, there being no other Julia TOON library.
    toons = Vector{Union{String,Nothing}}(undef, length(files))
    for (i, f) in enumerate(files)
        toons[i] = try
            CToon.dumps(CToon.parse(f.json))
        catch
            nothing
        end
    end

    ok = 0
    bytes = 0.0
    t0 = time()
    for r in 1:REPEATS
        for (i, f) in enumerate(files)
            t = toons[i]
            t === nothing && continue
            try
                CToon.to_json(CToon.parse_toon(t))
                ok += 1
                bytes += sizeof(t)
            catch e
                r == 1 && log_fail("toon_to_json", f.path, e)
            end
        end
    end
    return time() - t0, ok, bytes
end

function bench_roundtrip(files, log_fail)
    ok = 0
    bytes = 0.0
    t0 = time()
    for r in 1:REPEATS
        for f in files
            try
                toon = CToon.dumps(CToon.parse(f.json))
                CToon.to_json(CToon.parse_toon(toon))
                ok += 1
                bytes += sizeof(f.json)
            catch e
                r == 1 && log_fail("roundtrip", f.path, e)
            end
        end
    end
    return time() - t0, ok, bytes
end

function main()
    if length(ARGS) < 2
        println(stderr, "usage: julia bench.jl <manifest> <results.json> [log_file]")
        exit(1)
    end
    manifest_path, results_path = ARGS[1], ARGS[2]
    log_path = length(ARGS) >= 3 ? ARGS[3] : ""
    log_io = isempty(log_path) ? nothing : open(log_path, "w")

    log_fail = (op, path, err) -> (log_io === nothing || println(log_io, "[ctoon] [$op] FILE: $path ERROR: $err"))

    files = load_corpus(manifest_path)
    if isempty(files)
        println(stderr, "Corpus manifest is empty or unreadable.")
        exit(1)
    end

    total_json_bytes = sum(sizeof(f.json) for f in files)

    println("CToon Benchmarks — Julia")
    println("Corpus: $(length(files)) files, $(round(total_json_bytes / 1e6; digits=2)) MB (JSON)\n")
    println(rpad("Library", 8), " ", rpad("Operation", 14), " ", lpad("Throughput", 12), " ",
            lpad("Docs/sec", 14), " ", lpad("Success", 9), "  ", lpad("Total time", 12))

    results = Result[]

    for (op, bench_fn) in (
        ("json_to_toon", bench_json_to_toon),
        ("toon_to_json", bench_toon_to_json),
        ("roundtrip", bench_roundtrip),
    )
        seconds, ok, bytes = bench_fn(files, log_fail)
        record!(results, "ctoon", op, seconds, ok, length(files) * REPEATS, bytes)
    end

    write_results_json(results_path, "julia", length(files), total_json_bytes, results)
    println("\nResults written to $results_path")
    if log_io !== nothing
        println("Debug log written to $log_path")
        close(log_io)
    end
end

main()
